tools/crm_shadow.c

```diff
@@ -93,453 +145,568 @@ shadow_teardown(char *name)
+                            editor, shadow_file, strerror(errno));
+                goto done;
+            }
+            break;
 
-        if (editor == NULL) {
-            fprintf(stderr, "No value for EDITOR defined\n");
-            exit_code = CRM_EX_NOT_CONFIGURED;
-            goto done;
-        }
+        case shadow_cmd_switch:
+            // Switch to the named shadow instance
+            shadow_setup(options.shadow, TRUE);
+            break;
 
-        execlp(editor, "--", shadow_file, NULL);
-        fprintf(stderr, "Could not invoke EDITOR (%s %s): %s\n",
-                editor, shadow_file, strerror(errno));
-        exit_code = CRM_EX_OSFILE;
-        goto done;
+        case shadow_cmd_display:
+            // Display the current shadow file contents
+            {
+                char *output_s = NULL;
+                xmlNode *output = filename2xml(shadow_file);
 
-    } else if (command == 's') {
-        shadow_setup(shadow, TRUE);
-        goto done;
+                output_s = dump_xml_formatted(output);
+                printf("%s", output_s);
 
-    } else if (command == 'p') {
-        /* display the current contents */
-        char *output_s = NULL;
-        xmlNode *output = filename2xml(shadow_file);
+                free(output_s);
+                free_xml(output);
+            }
+            break;
 
-        output_s = dump_xml_formatted(output);
-        printf("%s", output_s);
+        case shadow_cmd_diff:
+            // Diff the shadow file against the cluster
+            {
+                xmlNode *diff = NULL;
+                xmlNode *old_config = NULL;
+                xmlNode *new_config = filename2xml(shadow_file);
+
+                rc = real_cib->cmds->query(real_cib, NULL, &old_config,
+                                           options.cmd_options);
+                if (rc != pcmk_ok) {
+                    rc = pcmk_legacy2rc(rc);
+                    exit_code = pcmk_rc2exitc(rc);
+                    g_set_error(&error, PCMK__EXITC_ERROR, exit_code,
+                                "Could not query the CIB: %s", pcmk_rc_str(rc));
+                    goto done;
+                }
 
-        free(output_s);
-        free_xml(output);
+                xml_track_changes(new_config, NULL, new_config, false);
+                xml_calculate_changes(old_config, new_config);
 
-    } else if (command == 'd') {
-        /* diff against cluster */
-        xmlNode *diff = NULL;
-        xmlNode *old_config = NULL;
-        xmlNode *new_config = filename2xml(shadow_file);
+                diff = xml_create_patchset(0, old_config, new_config, NULL,
+                                           false);
 
-        rc = real_cib->cmds->query(real_cib, NULL, &old_config, command_options);
+                xml_log_changes(LOG_INFO, __func__, new_config);
+                xml_accept_changes(new_config);
+                if (diff != NULL) {
+                    // @TODO: Is this supposed to be an error?
```

Comment:
I think the point is to behave like diff (0 means no differences). Something like CRM_EX_DIGEST might be better (to distinguish from actual errors) but we should keep backward compatibility at this point.